Numpy Prerelease Failure

Just had a build which failed with the numpy prerelease:

```
=================================== FAILURES ===================================

_______________________ TestUfuncCoverage.test_coverage ________________________

self = <astropy.units.tests.test_quantity_ufuncs.TestUfuncCoverage object at 0x7f53072ec2e8>

    def test_coverage(self):

        all_np_ufuncs = set([ufunc for ufunc in np.core.umath.__dict__.values()

                             if type(ufunc) == np.ufunc])

    

        from .. import quantity_helper as qh

    

        all_q_ufuncs = (qh.UNSUPPORTED_UFUNCS |

                        set(qh.UFUNC_HELPERS.keys()))

    

>       assert all_np_ufuncs - all_q_ufuncs == set([])

E       assert {<ufunc 'float_power'>} == set()

E         Extra items in the left set:

E         <ufunc 'float_power'>

E         Use -v to get the full diff

astropy/units/tests/test_quantity_ufuncs.py:25: AssertionError

= 1 failed, 8228 passed, 437 skipped, 71 xfailed, 1 xpassed in 272.09 seconds ==
```

Maybe that's an upstream issue but it's more likely they added another ufunc that should be supported with quantities.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1763,6 +1763,8 @@ Bug Fixes
     on python2. This solves bugs in parsing coordinates involving strings too.
     [#5355]
 
+  - Add support ``np.float_power``, which is new in numpy 1.12. [#5480]
+
 - ``astropy.utils``
 
   - Fixed AttributeError when calling ``utils.misc.signal_number_to_name`` with

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1763,7 +1763,8 @@ Bug Fixes
     on python2. This solves bugs in parsing coordinates involving strings too.
     [#5355]
 
-  - Add support ``np.float_power``, which is new in numpy 1.12. [#5480]
+  - Ensure ``Quantity`` supports ``np.float_power``, which is new in numpy 1.12.
+    [#5480]
 
 - ``astropy.utils``
 

--- a/astropy/units/quantity.py
+++ b/astropy/units/quantity.py
@@ -385,29 +385,31 @@ class Quantity(np.ndarray):
                                         args[0].__class__.__name__,
                                         args[1].__class__.__name__))
 
-        # In the case of np.power, the unit itself needs to be modified by an
-        # amount that depends on one of the input values, so we need to treat
-        # this as a special case.
-        # TODO: find a better way to deal with this case
-        if function is np.power and result_unit != dimensionless_unscaled:
-
-            if units[1] is None:
-                p = args[1]
+        # In the case of np.power and np.float_power, the unit itself needs to
+        # be modified by an amount that depends on one of the input values,
+        # so we need to treat this as a special case.
+        # TODO: find a better way to deal with this.
+        if result_unit is False:
+            if units[0] is None or units[0] == dimensionless_unscaled:
+                result_unit = dimensionless_unscaled
             else:
-                p = args[1].to(dimensionless_unscaled).value
-
-            try:
-                result_unit = result_unit ** p
-            except ValueError as exc:
-                # Changing the unit does not work for, e.g., array-shaped
-                # power, but this is OK if we're (scaled) dimensionless.
-                try:
-                    converters[0] = units[0]._get_converter(
-                        dimensionless_unscaled)
-                except UnitConversionError:
-                    raise exc
+                if units[1] is None:
+                    p = args[1]
                 else:
-                    result_unit = dimensionless_unscaled
+                    p = args[1].to(dimensionless_unscaled).value
+
+                try:
+                    result_unit = units[0] ** p
+                except ValueError as exc:
+                    # Changing the unit does not work for, e.g., array-shaped
+                    # power, but this is OK if we're (scaled) dimensionless.
+                    try:
+                        converters[0] = units[0]._get_converter(
+                            dimensionless_unscaled)
+                    except UnitConversionError:
+                        raise exc
+                    else:
+                        result_unit = dimensionless_unscaled
 
         # We now prepare the output object
         if self is obj:

--- a/astropy/units/quantity_helper.py
+++ b/astropy/units/quantity_helper.py
@@ -206,17 +206,20 @@ UFUNC_HELPERS[np.true_divide] = helper_division
 
 
 def helper_power(f, unit1, unit2):
+    # TODO: find a better way to do this, currently need to signal that one
+    # still needs to raise power of unit1 in main code
     if unit2 is None:
-        return [None, None], _d(unit1)
+        return [None, None], False
 
-    # TODO: find a better way to do this, currently
-    # need to raise power of unit1 in main code
     try:
-        return [None, get_converter(unit2, dimensionless_unscaled)], _d(unit1)
+        return [None, get_converter(unit2, dimensionless_unscaled)], False
     except UnitsError:
         raise TypeError("Can only raise something to a dimensionless quantity")
 
 UFUNC_HELPERS[np.power] = helper_power
+# float_power was added in numpy 1.12
+if isinstance(getattr(np, 'float_power', None), np.ufunc):
+    UFUNC_HELPERS[np.float_power] = helper_power
 
 
 def helper_ldexp(f, unit1, unit2):

--- a/astropy/units/tests/test_quantity_ufuncs.py
+++ b/astropy/units/tests/test_quantity_ufuncs.py
@@ -299,6 +299,15 @@ class TestQuantityMathFuncs(object):
         assert np.all(np.power(np.arange(4.) * u.m, 0.) ==
                       1. * u.dimensionless_unscaled)
 
+    # float_power only introduced in numpy 1.12
+    @pytest.mark.skipif("not hasattr(np, 'float_power')")
+    def test_float_power_array(self):
+        assert np.all(np.float_power(np.array([1., 2., 3.]) * u.m, 3.)
+                      == np.array([1., 8., 27.]) * u.m ** 3)
+        # regression check on #1696
+        assert np.all(np.float_power(np.arange(4.) * u.m, 0.) ==
+                      1. * u.dimensionless_unscaled)
+
     @raises(ValueError)
     def test_power_array_array(self):
         np.power(4. * u.m, [2., 4.])

